### PR TITLE
Enable coverage reporting

### PR DESCRIPTION
**Pull request by michaelfig**
See COVERAGE.md for more information.

### CODE REVIEW

**Comment by michaelfig** [comment]
We now break the deadlock as described in https://github.com/Agoric/agoric-sdk/issues/527#issuecomment-841747122 by using https://github.com/standard-things/esm/pull/877

```
soil:t michael$ ls
main.cjs                module.js               package.json
soil:t michael$ cat main.cjs 
require = require("esm")(module);
const m = require("./module.js");
console.log(m.X);
soil:t michael$ cat module.js 
console.log('in module');
export const X = 1;
soil:t michael$ node module.js
in module
soil:t michael$ node main.cjs
/Users/michael/agoric/agoric-sdk/packages/t/module.js:1
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/michael/agoric/agoric-sdk/packages/t/module.js
require() of ES modules is not supported.
require() of /Users/michael/agoric/agoric-sdk/packages/t/module.js from /Users/michael/agoric/agoric-sdk/packages/t/main.cjs is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename module.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/michael/agoric/agoric-sdk/packages/t/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1080:13) {
  code: 'ERR_REQUIRE_ESM'
}
```

After the ESM patch in this PR:

```
soil:t michael$ node main.cjs
in module
1
```

File: patches/esm+3.2.25.patch
Comment (dckc) [inline patches/esm+3.2.25.patch:8]: How did you make this patch? Is your work available in some sort of some source form?

Comment (michaelfig) [reply patches/esm+3.2.25.patch:8]: > How did you make this patch? Is your work available in some sort of some source form?

I pushed it to agoric-labs/esm#1

**Comment by michaelfig** [comment]
This is about as good as coverage is going to get until the caveats in `COVERAGE.md` are lifted.

File: .github/workflows/test-all-packages.yml
Comment (warner) [inline .github/workflows/test-all-packages.yml]: Does this give commit access to the `nwtgck` action, and does that expose us to Docker vulnerabilities of the sort that recently hit codecov? What authority does this action actually need? Is there a way to limit this somewhat? Maybe https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token ?

File: COVERAGE.md
Comment (warner) [inline COVERAGE.md]: That's a pretty big caveat. Do the percentages still seem to be accurate? Is there a human-manageable way to get from the line numbers this emits to the actual source code? Or should we preface all the reports with a note that we should ignore the line numbers entirely and just look at the percentages?

Comment (warner) [reply COVERAGE.md]: Also, does this affect basic unit tests that use `-r esm` (like everything we do) but do not use `bundle-source` or `import-bundle`? I'd be happy with getting accurate coverage for library code that we can test outside the context of swingset.

File: package.json
Comment (warner) [inline package.json:88]: Does this interfere with / override `ava` settings in individual package's `yarn test`?

It looks like this is meant for use by a top-level `yarn test` which runs tests in all packages in parallel. When I last looked into doing that, I got stuck on the different `require` options we needed for different sets of tests (in particular importing `install-ses`, or `install-metering-and-ses`). Did you find a solution for that?

File: packages/swing-store-simple/package.json
Comment (warner) [inline packages/swing-store-simple/package.json:8]: Ooh, does this work? I'm suspicious that this PR changes the type for this one package and none of the others. @kriskowal does this interact well/badly with the other ESM/`-r ESM` changes you're planning?

File: patches/esm+3.2.25.patch
Comment (warner) [inline patches/esm+3.2.25.patch:2]: I'm surprised this patch changed, since we didn't change the version of `esm` that we're using. Is this intentional? Same for the other patch.

File: packages/swing-store-simple/package.json
Comment (dckc) [reply packages/swing-store-simple/package.json:8]: Does https://github.com/Agoric/agoric-sdk/issues/527#issuecomment-841747122 answer these questions?

File: .github/workflows/test-all-packages.yml
Comment (michaelfig) [reply .github/workflows/test-all-packages.yml]: Thanks for the attention to this.  I couldn't figure out how to limit authority, so I've disabled access to the `GITHUB_TOKEN`.

File: COVERAGE.md
Comment (michaelfig) [reply COVERAGE.md]: So the basic plan is: if we adopt the Node.js `esm` package patch I made in this PR, we can do a gradual bottom-up migration of our packages to be dual `-r esm` and Node.js ESM (`nesm`) compatible.

I've changed the HTML file generation only to coverage-test the packages which can support nesm and explicitly implement a `test:c8` script.  Currently only `swing-store-simple` implements that (and incidentally, also supports `resm`).

File: package.json
Comment (michaelfig) [reply package.json:88]: > Does this interfere with / override `ava` settings in individual package's `yarn test`?

Only when run from the top level.  If `yarn test` is run in the specific package, the package's settings will take effect.

> It looks like this is meant for use by a top-level `yarn test` which runs tests in all packages in parallel. When I last looked into doing that, I got stuck on the different `require` options we needed for different sets of tests (in particular importing `install-ses`, or `install-metering-and-ses`). Did you find a solution for that?

Yes.  It turns out we only use `"require"` for two modules: `"esm"` (which appears in basically every `"ava"` config), and an enzyme setup script (only used in `packages/ui-components/package.json`).  So, the situation is not nearly as grim.

File: packages/swing-store-simple/package.json
Comment (michaelfig) [reply packages/swing-store-simple/package.json:8]: > Ooh, does this work?

Yep!

> I'm suspicious that this PR changes the type for this one package and none of the others. 

I did that to demonstrate that incremental, bottom-up support of NESM is possible without having to throw away `-r esm`.

> @kriskowal does this interact well/badly with the other ESM/`-r ESM` changes you're planning?

In talking with @kriskowal, this is a fine step forward.

File: patches/esm+3.2.25.patch
Comment (michaelfig) [reply patches/esm+3.2.25.patch:2]: > I'm surprised this patch changed, since we didn't change the version of `esm` that we're using. Is this intentional?

Yes, as discussed in https://github.com/Agoric/agoric-sdk/issues/527#issuecomment-841747122

> Same for the other patch.

`nyc` is no longer referenced, so I needed to remove that patch or `patch-package` complains about its absence.

**Review by katelynsills — COMMENTED** [review]
The changes to Zoe and ERTP look fine to me. I defer to whatever @warner and @dckc think, but let me know if there's anything you want my eye on.

File: COVERAGE.md
Comment (kriskowal) [inline COVERAGE.md]: Let’s spell this “Node.js ESM” formally.

**Review by kriskowal — APPROVED** [review]
It might be good to also change `scripts/repackage.sh` to reflect this new mode.

**Comment by kriskowal** [comment]
How is your fork of `esm` linked in this? I see nothing changed with the package.json name. Are we publishing as `esm` as if upstream.

**Comment by michaelfig** [comment]
> How is your fork of esm linked in this? I see nothing changed with the package.json name. Are we publishing as esm as if upstream.

We're patching upstream esm with `agoric-sdk/patches/esm+3.2.25.diff`

File: COVERAGE.md
Comment (michaelfig) [reply COVERAGE.md]: Done.

File: package.json
Comment (warner) [reply package.json:88]: Ok, I think @dtribble was once keen on using the AVA config to jam a `-r @agoric/install-ses` into e.g. all zoe tests, to avoid adding a line to the top of every test file, and that was going to push us to e.g. move all `@agoric/install-metering-and-ses` tests into one directory, and have two top-level AVA calls (with one set of `-r` options each). I think we talked him down from that :).

I've got `nodeArguments: [ "--expose-gc" ]` in the SwingSet package.json, to enable GC controls while testing. We may need to move that to the top-level package.json to make sure swingset tests work both from the top level and from `packages/SwingSet` (and also when e.g. Zoe does swingset-based tests, although if they're short enough, we can probably tolerate the inability to provoke GC).

File: .github/workflows/test-all-packages.yml
Comment (warner) [inline .github/workflows/test-all-packages.yml:168]: thanks!

File: COVERAGE.md
Comment (warner) [inline COVERAGE.md]: might remove the trailing comma, since that'll get rejected as non-JSON

**Review by warner — APPROVED** [review]
Ok, as long the top-level `yarn test` provides the right set of AVA options to the right packages (I'm specifically looking at `nodeArguments: [ "--expose-gc" ]` for SwingSet, as I'm about to add a test that depends upon it), I'm in.

**Pull request merged**
